File: Exercise_2.cpp
// Time Complexity                              :  O(N)
// Space Complexity                             :  O(N) asymptotically. Maximum would be the maximums of all the 2*2^h from the root. 
//                                                 Since we are using a local vector also in this case it will add to the space complexity.
// Did this code successfully run on Leetcode   :  Yes
// Any problem you faced while coding this      :  No

#include <bits/stdc++.h> 
using namespace std;  

// https://leetcode.com/problems/binary-tree-zigzag-level-order-traversal/submissions/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> ans;
        
        if(!root) return ans;
        
        queue<TreeNode*> q;
        q.push(root);
        bool reverseVector = false;
        while(!q.empty()) {
            int sz = q.size();
            vector<int> localVector;
            for(int i=0;i<sz;i++) {
                TreeNode *ctop = q.front();
                q.pop();
                localVector.push_back(ctop->val);
                if(ctop->left) q.push(ctop->left);
                if(ctop->right) q.push(ctop->right);
            }
            
            if(reverseVector == true) {
                reverse(begin(localVector),end(localVector));
                reverseVector = false;
            } else {
                reverseVector = true;
            }
            
            ans.push_back(localVector);
        }
        
        return ans;
    }
};

File: Exercise_2.py
#Time Complexity : O(N)), N being the number of nodes in the tree. 
#Space Complexity : O(logN), N being the number of nodes in the tree. 
#Did this code successfully run on Leetcode : Yes.
#Any problem you faced while coding this : Don't remember but probably had to debug a bit at the 
#end
#Your code here along with comments explaining your approach in three sentences only
'''Using DFS, for each level, create an array keep track of the largest. 
'''
class Solution:
    def helper(self, node, largest, level):
        if node!=None:
            #if there is no value for level, create with value smaller than smallest value. 
            if len(largest)<=level:
                largest.append(-2**32)
            #replace with value of node if value of node is larger. 
            if largest[level]<node.val:
                largest[level]=node.val
            #recurse.
            self.helper(node.left, largest, level+1)
            self.helper(node.right, largest, level+1)

    def largestValues(self, root: Optional[TreeNode]) -> List[int]:
        #initalize and call helper function. 
        largest = []
        level = 0
        if root!=None:
            self.helper(root, largest, level)
        return largest